src/Strings.h

```diff
@@ -0,0 +1,388 @@
+
+    return wordIterVector;
+}
+
+template <
+    typename IteratorT,
+    typename IsDelimiterT
+>
+inline void Parse(IteratorT const &begin,
+                  IteratorT const &end,
+                  IsDelimiterT const &isDelimiter,
+                  std::function<void (IteratorT&, IteratorT&)> const &callback) {
+
+    IteratorT left = begin;
+    IteratorT right = left;
+    while (left != end && *left == isDelimiter)
+        ++left;
+
+    right = left;
+    while (right != end) {
+        if (*right == isDelimiter) {
+            callback(left, right);
+            left = right;
+            while (left != end && *left == isDelimiter)
+                ++left;
+            right = left;
+        } else {
+            ++right;
+        }
+    }
+    if (left != right)
+        callback(left, right);
+}
+
+template <
+    typename IteratorT,
+    typename RegexT
+>
+inline void ParseRegex(IteratorT const &begin,
+                       IteratorT const &end,
+                       RegexT const &regexToken,
+                       std::function<void (IteratorT&, IteratorT&)> const &callback) {
+
+    std::sregex_iterator iter(begin, end, regexToken);
+    std::sregex_iterator iterEnd;
+
+    while (iter != iterEnd) {
+        std::string const iterStringConst = iter->str();
+        IteratorT beginRef = iterStringConst.begin();
+        IteratorT endRef = iterStringConst.end();
+        callback(beginRef, endRef);
+        ++iter;
+    }
+}
+
+template <typename IteratorT>
+inline std::string TrimAndReplace(IteratorT begin,
+                                  IteratorT end) {
+    //replace all punctuations with spaces
+    std::transform(begin, end, begin, [](char c) {
+        if (std::ispunct(c) )
+            return ' ';
+        return c;
+    });
+
+    std::string result(begin, end);
+    //remove duplicate space
+    result.erase(
+        std::unique(
+            result.begin(),
+            result.end(),
+            [=](char lhs, char rhs){
+                return (lhs == rhs) && (lhs == ' ');
+            }
+        ),
+        result.end()
+    );
+
+    return result;
+}
+
+template <typename IteratorT>
+inline void ParseNgramCharHelper(IteratorT const &begin,
+                                 IteratorT const &end,
+                                 size_t const &ngramRangeMin,
+                                 size_t const &ngramRangeMax,
+                                 std::function<void (IteratorT&, IteratorT&)> const &callback) {
+
+    std::vector<IteratorT> wordIterVector(IteratorVectorGenerator(begin, end));
+
+    for (size_t strIterOffset = 0; strIterOffset < wordIterVector.size(); ++strIterOffset)  {
+        for (size_t ngramRangeVal = ngramRangeMin; ngramRangeVal <= ngramRangeMax; ++ngramRangeVal) {
+            if (strIterOffset + ngramRangeVal < wordIterVector.size())
+                callback(wordIterVector[strIterOffset], wordIterVector[strIterOffset + ngramRangeVal]);
```

Comment:
>strIterOffset + ngramRangeVal [](start = 71, length = 29)

Should we assert that the resulting iterator does not point beyond the vector? Operator[] does not check bounds.